File: 2022/Day-20/tasks.rb
def solution(nums, coef = 1, cycles = 1)
  # [{ndx:0, val: 1}, {ndx: 1, val: 5},...]
  items = nums.map.with_index { |el, ndx| { ndx: ndx, val: el.to_i * coef } }

  cycles.times do
    nums.size.times do |ndx|
      before_ndx = items.find_index { |el| el[:ndx] == ndx }
      item = items.delete_at(before_ndx)
      after_ndx = (before_ndx + item[:val]) % items.size
      items.insert(after_ndx, item)
    end
  end

  zero_ndx = items.find_index { |el| el[:val] == 0 }

  [1000, 2000, 3000].sum do |offset| 
    items[(zero_ndx + offset) % nums.size][:val]
  end
end

numbers = File.readlines("input0.txt").map(&:to_i)
puts "Part1: ", solution(numbers)
puts "Part2: ", solution(numbers, 811589153, 10)

File: 2022/Day-20/tasks.js
// OPEN GPT rewrote it for me. Just for curiosity and for 
function solution(nums, coef = 1, cycles = 1) {
  let items = nums.map((el, ndx) => ({ ndx, val: el * coef }));

  for (let i = 0; i < cycles; i++) {
    for (let ndx = 0; ndx < nums.length; ndx++) {
      let before = items.findIndex(el => el.ndx === ndx);
      let item = items.splice(before, 1)[0];
      let after = (before + item.val) % items.length;
      if (after < 0) after += items.length;
      items.splice(after, 0, item);
    }
  }

  let zeroNdx = items.findIndex(el => el.val === 0);
  let coords = [1000, 2000, 3000];
  return coords.reduce((sum, coord) => {
    return sum + items[(zeroNdx + coord) % nums.length].val;
  }, 0);
}

const numbers = require('fs').readFileSync('input1.txt', 'utf-8').split('\n').map(Number);
console.time("code");
console.log(`Part1: ${solution(numbers)}`);
console.log(`Part2: ${solution(numbers, 811589153, 10)}`);
console.timeEnd("code");
